### PR TITLE
fix: Clone the configuration object. If the name of the configuration object conflicts, the clone can still be successful (#334)

### PR DESCRIPTION
**Pull request by MapleWithoutWords**
fix:Clone the configuration object. If the name of the configuration object conflicts, the clone can still be successful.
image
image

### CODE REVIEW

File: src/Services/Masa.Dcc.Service/Domain/App/Repositories/IAppConfigObjectRepository.cs [hunk 1]
```diff
@@ -5,6 +5,7 @@ namespace Masa.Dcc.Service.Admin.Domain.App.Repositories
 {
     public interface IAppConfigObjectRepository : IRepository<AppConfigObject>
     {
+        Task<List<AppConfigObject>> GetListByAppIdAsync(int appId);
```
Comment (doddgu) [inline src/Services/Masa.Dcc.Service/Domain/App/Repositories/IAppConfigObjectRepository.cs:8]: Add blank line in the middle of methods.

File: src/Services/Masa.Dcc.Service/Domain/App/Repositories/IBizConfigObjectRepository.cs [hunk 2]
```diff
@@ -6,5 +6,6 @@ namespace Masa.Dcc.Service.Admin.Domain.App.Repositories
     public interface IBizConfigObjectRepository : IRepository<BizConfigObject>
     {
         Task<List<BizConfigObject>> GetListByEnvClusterIdAsync(int envClusterId, int bizConfigId);
+        Task<List<BizConfigObject>> GetListByBizConfigIdAsync(int bizConfigId);
```
Comment (doddgu) [inline src/Services/Masa.Dcc.Service/Domain/App/Repositories/IBizConfigObjectRepository.cs:9]: Add blank line in the middle of methods.

File: src/Services/Masa.Dcc.Service/Domain/App/Repositories/IPublicConfigObjectRepository.cs [hunk 3]
```diff
@@ -8,5 +8,6 @@ public interface IPublicConfigObjectRepository : IRepository<PublicConfigObject>
         Task<List<PublicConfigObject>> GetListByEnvClusterIdAsync(int? envClusterId, int publicConfigId);
 
         Task<PublicConfigObject> GetByConfigObjectIdAsync(int configObjectId);
+        Task<List<PublicConfigObject>> GetListByPublicConfigIdAsync(int publicConfigId);
```
Comment (doddgu) [inline src/Services/Masa.Dcc.Service/Domain/App/Repositories/IPublicConfigObjectRepository.cs:11]: Add blank line in the middle of methods.

File: src/Services/Masa.Dcc.Service/Domain/App/Services/ConfigObjectDomainService.cs [hunk 4]
```diff
@@ -225,6 +227,63 @@ private async Task CloneConfigObjectsAsync(List<AddConfigObjectDto> configObject
...
+        /// cloning, check config object whether already exists.
+        /// </summary>
+        /// <param name="configObjects"></param>
+        /// <param name="appId">the 'appId' argument may be is 'application id' or 'public config Id' or 'biz config id'</param>
+        /// <returns></returns>
+        /// <exception cref="UserFriendlyException"></exception>
+        private async Task CheckConfigObjectDuplication(List<AddConfigObjectDto> configObjects, int appId)
+        {
+            if (configObjects != null && configObjects.Count > 0)
+
```
Comment (doddgu) [inline src/Services/Masa.Dcc.Service/Domain/App/Services/ConfigObjectDomainService.cs:240]: Remove the blank line.

File: src/Services/Masa.Dcc.Service/Domain/App/Services/ConfigObjectDomainService.cs [hunk 5]
```diff
@@ -225,6 +227,63 @@ private async Task CloneConfigObjectsAsync(List<AddConfigObjectDto> configObject
...
+        /// <summary>
+        /// cloning, check config object whether already exists.
+        /// </summary>
+        /// <param name="configObjects"></param>
+        /// <param name="appId">the 'appId' argument may be is 'application id' or 'public config Id' or 'biz config id'</param>
+        /// <returns></returns>
+        /// <exception cref="UserFriendlyException"></exception>
+        private async Task CheckConfigObjectDuplication(List<AddConfigObjectDto> configObjects, int appId)
+        {
+            if (configObjects != null && configObjects.Count > 0)
```
Comment (doddgu) [inline src/Services/Masa.Dcc.Service/Domain/App/Services/ConfigObjectDomainService.cs:239]: `configObjects?.Count > 0`

File: src/Services/Masa.Dcc.Service/Domain/App/Services/ConfigObjectDomainService.cs [hunk 6]
```diff
@@ -225,6 +227,63 @@ private async Task CloneConfigObjectsAsync(List<AddConfigObjectDto> configObject
             await _configObjectRepository.AddRangeAsync(cloneConfigObjects);
         }
 
+        /// <summary>
+        /// cloning, check config object whether already exists.
```
Comment (doddgu) [inline src/Services/Masa.Dcc.Service/Domain/App/Services/ConfigObjectDomainService.cs:231]: Comment is `already exists`, and the method name is `duplication`?

File: src/Services/Masa.Dcc.Service/Domain/App/Services/ConfigObjectDomainService.cs [hunk 7]
```diff
@@ -225,6 +227,63 @@ private async Task CloneConfigObjectsAsync(List<AddConfigObjectDto> configObject
...
+        /// </summary>
+        /// <param name="configObjects"></param>
+        /// <param name="appId">the 'appId' argument may be is 'application id' or 'public config Id' or 'biz config id'</param>
```
Comment (doddgu) [inline src/Services/Masa.Dcc.Service/Domain/App/Services/ConfigObjectDomainService.cs:234]: Why appId can be or many times.

File: src/Services/Masa.Dcc.Service/Domain/App/Services/ConfigObjectDomainService.cs [hunk 8]
```diff
@@ -225,6 +227,63 @@ private async Task CloneConfigObjectsAsync(List<AddConfigObjectDto> configObject
...
+                                throw new UserFriendlyException($"Configuration Name '{item.Name}' already exist in the environment cluster '{item.EnvironmentClusterId}'.");
+                            }
+                        }
+                        break;
+                    case ConfigObjectType.Biz:
+                        var bizConfigObjects = await _bizConfigObjectRepository.GetListByBizConfigIdAsync(appId);
+                        foreach (var item in configObjects)
+                        {
+                            if (bizConfigObjects.Any(e => e.EnvironmentClusterId == item.EnvironmentClusterId
+                                                                    && e.ConfigObject.Name == item.Name))
```
Comment (doddgu) [inline src/Services/Masa.Dcc.Service/Domain/App/Services/ConfigObjectDomainService.cs:263]: Format code.

File: src/Services/Masa.Dcc.Service/Domain/App/Services/ConfigObjectDomainService.cs [hunk 9]
```diff
@@ -225,6 +227,63 @@ private async Task CloneConfigObjectsAsync(List<AddConfigObjectDto> configObject
...
+                switch (configType)
+                {
+                    case ConfigObjectType.Public:
+                        var allPublicConfigs = await _publicConfigObjectRepository.GetListByPublicConfigIdAsync(appId);
+                        foreach (var item in configObjects)
+                        {
+                            if (allPublicConfigs.Any(e => e.EnvironmentClusterId == item.EnvironmentClusterId
+                                                                    && e.ConfigObject.Name == item.Name))
+                            {
+
```
Comment (doddgu) [inline src/Services/Masa.Dcc.Service/Domain/App/Services/ConfigObjectDomainService.cs:253]: Remove the blank line.

File: src/Services/Masa.Dcc.Service/Domain/App/Services/ConfigObjectDomainService.cs [hunk 10]
```diff
@@ -225,6 +227,63 @@ private async Task CloneConfigObjectsAsync(List<AddConfigObjectDto> configObject
...
+                        }
+                        break;
+                    case ConfigObjectType.Biz:
+                        var bizConfigObjects = await _bizConfigObjectRepository.GetListByBizConfigIdAsync(appId);
+                        foreach (var item in configObjects)
+                        {
+                            if (bizConfigObjects.Any(e => e.EnvironmentClusterId == item.EnvironmentClusterId
+                                                                    && e.ConfigObject.Name == item.Name))
+                            {
+
```
Comment (doddgu) [inline src/Services/Masa.Dcc.Service/Domain/App/Services/ConfigObjectDomainService.cs:265]: Remove the blank line.

File: src/Services/Masa.Dcc.Service/Domain/App/Services/ConfigObjectDomainService.cs [hunk 11]
```diff
@@ -225,6 +227,63 @@ private async Task CloneConfigObjectsAsync(List<AddConfigObjectDto> configObject
...
+                        }
+                        break;
+                    case ConfigObjectType.App:
+                        var allAppConfigObjects = await _appConfigObjectRepository.GetListByAppIdAsync(appId);
+                        foreach (var item in configObjects)
+                        {
+                            if (allAppConfigObjects.Any(e => e.EnvironmentClusterId == item.EnvironmentClusterId
+                                                                    && e.ConfigObject.Name == item.Name))
+                            {
+
```
Comment (doddgu) [inline src/Services/Masa.Dcc.Service/Domain/App/Services/ConfigObjectDomainService.cs:277]: Remove the blank line.

File: src/Services/Masa.Dcc.Service/Infrastructure/Repositories/App/PublicConfigObjectRepository.cs [hunk 12]
```diff
@@ -30,5 +30,14 @@ public async Task<PublicConfigObject> GetByConfigObjectIdAsync(int configObjectI
 
             return result ?? new(0, 0);
         }
+        public async Task<List<PublicConfigObject>> GetListByPublicConfigIdAsync(int publicConfigId)
```
Comment (doddgu) [inline src/Services/Masa.Dcc.Service/Infrastructure/Repositories/App/PublicConfigObjectRepository.cs:33]: Add blank line.

**Review by doddgu — CHANGES_REQUESTED** [review]

File: src/Services/Masa.Dcc.Service/Domain/App/Services/ConfigObjectDomainService.cs (hunk 7 above)
Comment (MapleWithoutWords) [reply src/Services/Masa.Dcc.Service/Domain/App/Services/ConfigObjectDomainService.cs:234]: @doddgu  the clone method is general. it's may be clone app's configuration or public's configuration or biz's configuration. so the appId is many ppssibilities in Check method

**Pull request closed without merging**